accumulator/batch.go

```diff
@@ -0,0 +1,249 @@
+	// request ID as the key.
+	invocations                 map[string]*Invocation
+	count                       int
+	age                         time.Time
+	maxSize                     int
+	maxAge                      time.Duration
+	currentlyExecutingRequestID string
+
+	// TODO: @lahsivjar remove requirements of a mutex; currently it is
+	// required because the invocations need to be accessed from logsapi
+	// as the processed log output of logsapi doesn't have the necessary
+	// information.
+	mu sync.RWMutex
+}
+
+// NewBatch creates a new BatchData which can accept a
+// maximum number of entries as specified by the arguments.
+func NewBatch(maxSize int, maxAge time.Duration) *Batch {
+	return &Batch{
+		invocations: make(map[string]*Invocation),
+		maxSize:     maxSize,
+		maxAge:      maxAge,
+	}
+}
+
+// RegisterInvocation registers a new function invocation against its request
+// ID. It also updates the caches for currently executing request ID.
+func (b *Batch) RegisterInvocation(
+	requestID, functionARN string,
+	deadlineMs int64,
+	timestamp time.Time,
+) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.invocations[requestID] = &Invocation{
+		RequestID:   requestID,
+		FunctionARN: functionARN,
+		DeadlineMs:  deadlineMs,
+		Timestamp:   timestamp,
+	}
+	b.currentlyExecutingRequestID = requestID
+}
+
+// OnAgentInit caches the transactionID and the payload for the currently
+// executing invocation as reported by the agent. The traceID and transactionID
+// will be used to create a new transaction in an event the actual transaction
+// is not reported by the agent due to unexpected termination.
+func (b *Batch) OnAgentInit(transactionID string, traceID string) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	i, ok := b.invocations[b.currentlyExecutingRequestID]
+	if !ok {
+		return fmt.Errorf("invocation for requestID %s does not exist", b.currentlyExecutingRequestID)
+	}
+	i.TransactionID, i.TraceID = transactionID, traceID
+	return nil
+}
+
+// AddAgentData adds a data received from agent. For a specific invocation
+// agent data is always received in the same invocation.
+func (b *Batch) AddAgentData(apmData APMData) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	inc, ok := b.invocations[b.currentlyExecutingRequestID]
+	if !ok {
+		return fmt.Errorf("invocation for requestID %s does not exist", b.currentlyExecutingRequestID)
+	}
+
+	raw, err := GetUncompressedBytes(apmData.Data, apmData.ContentEncoding)
+	if err != nil {
+		return err
+	}
+	// A request body can either be empty or have a ndjson content with
+	// first line being metadata.
+	data := bytes.Split(raw, []byte("\n"))
+	if len(data) == 0 {
+		return nil
+	}
+	// Set metadata if not already set
+	if b.metadataBytes == 0 {
+		b.metadataBytes, _ = b.buf.Write(data[0])
+	}
+	// Should we identify the target transaction and cache only that?
+	for i := 1; i < len(data); i++ {
+		if d := data[i]; len(d) > 0 {
+			inc.agentData = append(inc.agentData, d)
+		}
+	}
```

Comment:
> That seems sensible to me, since we're not enriching the events in the extension. 

Does the APM-Server and ES have any constraints around the order of events. I was thinking that if a transaction event is not reported for an invocation but other events are reported then will shipping the other events before the txn cause an unexpected state? (In this case a proxy transaction will be created and will be shipped at a later time but other events for the same invocation might end up earlier)

> That would imply that we would revert to sending the events through to APM Server immediately, right?

Since we will need to read the body of the payload anyway, I was thinking to have a single batch that will be shipped to APM-Server instead of shipping agent data individually. The batch will have a metadata and a combination of lambda events + agents events. The batch will be shipped to APM Server according to the batch maturity settings in the extension (age, size, shutdown event). In this case, for agent payload, the extension will:

1. Read the payload from agents line by line and strip metadata
2. Read the type of event on each line; if the event is the target transaction we will mark the invocation as finalized otherwise we will wait for the `runtimeDone` event from lambda logs (depending on the constraints of the order of events this can also be optimized).

WDYT?

Reply:
> Does the APM-Server and ES have any constraints around the order of events.

Transactions do not have to precede the other events. In general they won't; enclosed events (such as spans) will typically be ended before their enclosing transaction, and thus be sent to APM Server first.

> Since we will need to read the body of the payload anyway, I was thinking to have a single batch that will be shipped to APM-Server instead of shipping agent data individually. The batch will have a metadata and a combination of lambda events + agents events.

OK, I think that sounds good.

Reply:
I have updated the code to not cache any agent data. The logic checks if it has seen the root txn or not; in both cases it appends the event to the batch. If the txn is not seen until shutdown or `runtimeDone` for that invocation, we create a proxy txn.

Reply:
Looks good!